Fixed Mini BO avatar images not deleting when updating images

db306 opened the pull request:
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | When updating an object that inherits the ObjectModel with a new image, the cached thumbnail generated by the HelperList does not update as it doesn't get deleted properly.
| Type?         | bug fix
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes [BOOM-5503](http://forge.prestashop.com/browse/BOOM-5503) ticket
| How to test?  | Requires a custom implementation of an ObjectModel that uses the HelperList Class and displays an image.

When displaying images of a custom ObjectModel in a list using the **HelperList** Class within the BO, it generates a thumbnail of 45px that is later on displayed on the table. This image generation is triggered on : `classes/helper/HelperList.php` on line **312** : 

```
$this->_list[$index][$key] = ImageManager::thumbnail($path_to_image, $this->table.'_mini_'.$item_id.'_'.$this->context->shop->id.'.'.$this->imageType, 45, $this->imageType);
```
When updating an image of a custom ObjectModel implementation, it triggers the function deleteImage from the ObjectModel class that deletes existing images:

**/classes/ObjectModel.php** lines **1753** to **1769**

```
if ($this->image_dir) {
    if (file_exists($this->image_dir.$this->id.'.'.$this->image_format)
        && !unlink($this->image_dir.$this->id.'.'.$this->image_format)) {
        return false;
    }
}
if (file_exists(_PS_TMP_IMG_DIR_.$this->def['table'].'_'.$this->id.'.'.$this->image_format)
    && !unlink(_PS_TMP_IMG_DIR_.$this->def['table'].'_'.$this->id.'.'.$this->image_format)) {
    return false;
}
if (file_exists(_PS_TMP_IMG_DIR_.$this->def['table'].'_mini_'.$this->id.'.'.$this->image_format)
    && !unlink(_PS_TMP_IMG_DIR_.$this->def['table'].'_mini_'.$this->id.'.'.$this->image_format)) {
    return false;
}
```

The former piece of code will create an image in a path structure such as `/img/tmp/customtable_mini_15_1.jpg` whereas the equivalent example on the later piece of code will delete an image in a uri `/img/tmp/customtable_mini_15.jpg` given that these urls are not the same, when updating an image it will not delete its "cached" thumbnail.

Concerning the PR I have added this piece of code that checks whether the ObjectModel supports Multishop and then iterates through all shop ids or a single shop id depending on the current view. Then deletes existing thumbnail images.

Another way, and **probably a best approach** would be to change the HelperList thumbnail generation to stop using the shop_id as a variable in its path url. I assumed this had a reason to be here and hence I have changed the ObjectModel function instead. Let me know and I can modify my PR to change the thumbnail url instead :

This :
```
ImageManager::thumbnail($path_to_image, $this->table.'_mini_'.$item_id.'.'.$this->imageType, 45, $this->imageType);
```

instead of currently this:
```
ImageManager::thumbnail($path_to_image, $this->table.'_mini_'.$item_id.'_'.$this->context->shop->id.'.'.$this->imageType, 45, $this->imageType);
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/9045)
<!-- Reviewable:end -->

db306 committed:
Fixed Mini BO avatar images not deleting when updating images

db306 commented:
`description` isn't empty. Or is it?

PierreRambaud commented on classes/ObjectModel.php:
```diff
@@ -1767,6 +1767,20 @@ public function deleteImage($force_delete = false)
                 && !unlink(_PS_TMP_IMG_DIR_.$this->def['table'].'_mini_'.$this->id.'.'.$this->image_format)) {
                 return false;
             }
+            if($this->isMultishop()){
```
Will be easier to read and remove duplicates :)

```php
if ($this->isMultishop()) {
    $ids = $this->id_shop;
    if (!is_array($ids)) {
        $ids = [$ids];
    }

    foreach ($ids as $idShop) {
        $filePath = _PS_TMP_IMG_DIR_ .
                  $this->def['table'] .
                  '_mini_' .
                  $this->id .
                  '_' .
                  $idShop .
                  '.' .
                  $this->image_format;
        if (file_exists($filePath) && !unlink($filePath) {
            return false;
        }
    }
}
```

db306 replied:
👍

db306 committed:
Simplified code and removed duplicates

db306 committed:
Fixed spacing following Codacy feedback

PierreRambaud commented on classes/ObjectModel.php:
```diff
@@ -1767,15 +1767,15 @@ public function deleteImage($force_delete = false)
-            if($this->isMultishop()){
+            if ($this->isMultishop()){
 
                 $ids_shop = $this->id_shop;
```
Could be better with camelcase instead of underscore variable :)

db306 replied:
sure

PierreRambaud commented on classes/ObjectModel.php:
```diff
@@ -1767,15 +1767,15 @@ public function deleteImage($force_delete = false)
-                    if(file_exists(_PS_TMP_IMG_DIR_.$this->def['table'].'_mini_'.$this->id.'_'.$id_shop.'.'.$this->image_format)
-                        && !unlink(_PS_TMP_IMG_DIR_.$this->def['table'].'_mini_'.$this->id.'_'.$id_shop.'.'.$this->image_format)){
+                foreach ($ids_shop as $id_shop) {
+                    if (file_exists(_PS_TMP_IMG_DIR_.$this->def['table'].'_mini_'.$this->id.'_'.$id_shop.'.'.$this->image_format)
```
Can you please put file path into a variable to remove duplicate concatenation?

db306 replied:
I would heave done it that way too so I cannot agree more, but the whole function is built with double concatenation. I'll change them all and give them meaningful variable names.

PierreRambaud replied:
Nice, really easier to read now!

db306 committed:
Refactored var names and removed string concatenation duplicates

db306 committed:
Fixed spacings from latest Codacy review

jolelievre commented:
seems good to me!
@PierreRambaud ?

PierreRambaud commented:
Thanks @jolelievre forgot this one.
@db306 Can you send us a module or a script to test it? :)

db306 commented:
Thanks, I’ll look if I can make some tests. I’m leaving on vacations at the moment so it will have to wait til i get back. 

> On 21 Aug 2018, at 17:33, GoT <notifications@github.com> wrote:
> 
> Thanks @jolelievre forgot this one.
> @db306 Can you send us a module or a script to test it? :)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

sarjon committed:
init languages migrate

sarjon committed:
add languages grid definition factory

sarjon committed:
add language query builder

sarjon committed:
add languages field rendering

sarjon committed:
add sorting & pagination to language query builder

sarjon committed:
complete doc for controller

sarjon committed:
configure js

sarjon committed:
add links to legacy page

sarjon committed:
add decorated category data provider

sarjon committed:
remove logging

sarjon committed:
init customers list migration

sarjon committed:
add customer grid definition

sarjon committed:
add customers list rendering

sarjon committed:
add row actions

sarjon committed:
add new customer toolbar button

sarjon committed:
enable sidebar

sarjon committed:
add decorator for customers grid data factory

sarjon committed:
add filters definition for customer grid

sarjon committed:
update filters for customers

sarjon committed:
implementing filters

sarjon committed:
add filters to query builder

sarjon committed:
improve SearchAndResetType configuration

sarjon committed:
add js for customers

sarjon committed:
remove unused query builder param

sarjon committed:
add sorting for customers grid

sarjon committed:
add grid actions

sarjon committed:
add bulk actions

db306 commented:
Hello, it says the branch has conflicts that must be resolved ? I don't understand what is conflicting here

sarjon committed:
add total spent column formatting

sarjon committed:
add name to Edit action

sarjon committed:
cs fixer

sarjon committed:
reuse common controller for filtering/sorting

sarjon committed:
add missing docsblocks

sarjon committed:
update wording

sarjon committed:
fix rebase

sarjon committed:
add flag column to grid

sarjon committed:
add language flag thumbnail provider

sarjon committed:
add permissions check for languages list

sarjon committed:
cs fixer

sarjon committed:
fix languages sorting/filtering

sarjon committed:
init customers view

sarjon committed:
define contracts for customer information query

sarjon committed:
add personal customer information collecting

sarjon committed:
add customer order information

sarjon committed:
add customer carts information

sarjon committed:
add customer products data collecting

sarjon committed:
collect customer messages information

sarjon committed:
collect customer discounts data

sarjon committed:
collect information about last emails sent to customer

sarjon committed:
collect last connections of customer

sarjon committed:
add customer group information

sarjon committed:
add customer referrers information collecting

sarjon committed:
cs fixer

sarjon committed:
add customer adddress information collecting

sarjon committed:
remove redundant suffixes

sarjon committed:
cs fixer

sarjon committed:
start building view ui

PierreRambaud committed:
Try to change StockManagement Tests

sarjon committed:
typehint param

sarjon committed:
fix missing translations

sarjon committed:
Revert "Try to change StockManagement Tests"

This reverts commit 3e5840cb69ff69a9e754d63e6dc2e18399822eed.

sarjon committed:
add helper card

sarjon committed:
add helper card documentation link provider

sarjon committed:
remove card when close is clicked

sarjon committed:
add bg image for team helper card

sarjon committed:
update customer filtering by registration date

sarjon committed:
cs fixer

rokaszygmantas committed:
Prevent fatal error on session expiration in import step 2

rokaszygmantas committed:
Use POST when accessing import step 2 page

sarjon committed:
wip customer information display

Commit:
adds dots icon instead of arrow down on the new grid listing

Commit:
positions grid cog to be on the right of the list header

Commit:
adds card header title class which changes fonts

Commit:
modifies grid action extensions so it considerns now that all actions are located on grid header

Commit:
changes to display inline-block instead of inline-flex

Commit:
removes commented block

sarjon committed:
finish building personal information view block

sarjon committed:
add order information block

sarjon committed:
add carts information block

sarjon committed:
add bought products information block

sarjon committed:
add viewed products information

sarjon committed:
add customer addresses information block

sarjon committed:
add private note block

sarjon committed:
add messages information block

sarjon committed:
add discounts information block

sarjon committed:
add sent emails information block

sarjon committed:
add last connections information

sarjon committed:
add groups information display

sarjon committed:
update templates naming

sarjon committed:
minor fixes to customers view layout

sarjon committed:
fix listing of discounts

tomas862 committed:
fix CustomerException parent

Co-Authored-By: sarjon <jonusas.sarunas@gmail.com>

sarjon committed:
fix namespace for exception

sarjon committed:
minor ui changes

sarjon committed:
fix translations

sarjon committed:
add security and demo checks

sarjon committed:
handle customer not found exception

sarjon committed:
cs fixer

sarjon committed:
fix order date format

sarjon committed:
remove unused choices in customer grid definition

sarjon committed:
make ToggleColumn sortable

sarjon committed:
add sorting for status columns

sarjon committed:
update quotes for languages js file

rokaszygmantas committed:
Rework import form handler for both step 1 and step 2

rokaszygmantas committed:
Clean up import form handling

rokaszygmantas committed:
Fix cs

sarjon committed:
make documentation provider configurable

sarjon committed:
add update seo and url documnetation provider

sarjon committed:
remove helper doc service

sarjon committed:
minor code formatting fixes

sarjon committed:
build assets

sarjon committed:
fix typo

Commit:
Merge pull request #11249 from Seleda/fix4

Fix declaration of method arguments

Quetzacoalt91 committed:
$this->fetch() can return content in BO context

Commit:
Merge pull request #11261 from mickaelandrieu/fix-11141-inconsistent-locale-in-FO

Fixed inconsistent locale in Front Office

Commit:
Long namespace are not needed

Quetzacoalt91 committed:
Merge pull request #11058 from PierreRambaud/add-languages

Add Hindi, Mexican & Bosnian languages

Quetzacoalt91 committed:
Merge pull request #11179 from khouloudbelguith/fix-issues-11178-sort-ws

Fix sorting issue in the webservice page

Quetzacoalt91 committed:
Merge pull request #11215 from dineshbadrukhiya/Fix-11211

Fix 11211: Removing inline-style of color in test-data

Commit:
Fix cs

Commit:
TE: Add some fixes to the script

Quetzacoalt91 committed:
Merge pull request #11070 from PierreRambaud/feature/module-manager-categories

Feature/module manager categories

PierreRambaud committed:
Merge pull request #9316 from prestamodule/patch-107

Fix Cart::autoAddToCart using undefined customer when called from BO

PierreRambaud committed:
Merge pull request #9132 from PeeyushAgrawalWebkul/develop

extra field hook added on customer address form

PierreRambaud committed:
Merge pull request #11246 from MartinFinkenflugel/develop

Fix 10962: Replace deprecated library PhpExcel by PhpSpreadsheet

PierreRambaud committed:
Merge pull request #10904 from Matt75/issue-10876

[BO] Fix date filtering in AdminSpecificPriceRuleController

Commit:
Use the right collection number instead the children length

PierreRambaud committed:
Merge pull request #11266 from Quetzacoalt91/fetch-smarty-bo

$this->fetch() can return content in BO context

PierreRambaud committed:
Merge pull request #11214 from matks/fix-11180-maofree-fix-address

Handles bad address usecase for Carrier Tax Rate

Quetzacoalt91 committed:
Fix error on tab unregistration when handled by the module

sarjon committed:
build assets

Quetzacoalt91 committed:
Merge pull request #11287 from jolelievre/fix-filemanager-175

Fix filemanager security breaches

sarjon committed:
combine 2 services into 1

Quetzacoalt91 committed:
Fix linter

Quetzacoalt91 committed:
Merge pull request #11306 from Quetzacoalt91/lint-fix

Fix linter jobs of Travis

Quetzacoalt91 committed:
Merge pull request #11301 from PrestaShop/1.7.5.x

Merge 1.7.5.x in develop

jolelievre committed:
Add LegacyContext::getLegacyAdminLink because legacy url might be needed in some edge cases

Commit:
Be able to disable Help Link

jolelievre committed:
feedback from PR in js

jolelievre committed:
prevent short description limit to be set at 0

Commit:
Merge pull request #11300 from Quetzacoalt91/fix-deletion-of-parent-tab

Fix error on tab unregistration when handled by the module

Commit:
Merge pull request #11184 from ansar21mallouli/branch-language

Advanced localization

Quetzacoalt91 committed:
Merge pull request #10197 from mickaelandrieu/tools/schema-generator

Added a schema generator

PierreRambaud committed:
Merge pull request #11095 from tomas862/features/grid-wheel-icon-position-and-dropdown-toggle-modification

Introduces grid design update regarding grid wheel icon position and dropdown toggle icon modification

matks committed:
Merge pull request #10845 from sarjon/fix-services-naming

Minor services naming fix

jolelievre committed:
the twig helper is_ordering_column becomes more dynamic instead of checking hard coded position

sarjon committed:
build assets

jolelievre committed:
remove dump

matks committed:
Merge pull request #11049 from sarjon/migrate/employee-helper-card

Add Helper Card to Employees page

matks committed:
Merge pull request #10956 from PierreRambaud/fix-product-features

Fix product features

matks committed:
Merge pull request #11147 from jolelievre/position-ux-modif

PositionColummn UI/UX modifications

jolelievre committed:
allow all logs to be shown even when no employee is assigned

jolelievre committed:
Merge pull request #11312 from jolelievre/short-description-limit

Prevent short description limit to be set at 0

jolelievre committed:
translate N/A

jolelievre committed:
Manage redirection after email translation

Commit:
Update translations catalog

matks committed:
Only show warning install about latest stable version if version number < latest stable version

Quetzacoalt91 committed:
Merge pull request #11344 from jolelievre/emails-translation

Manage redirection after email translation

Quetzacoalt91 committed:
Merge pull request #11202 from jocel1/fix-cookie

fix a bug where PrestaShop cookie is set on every page when a cookie value is NULL

jolelievre committed:
transfer getLegacyAdminLink in the Link class

jolelievre committed:
remove useless use statements and cs fix Link

jolelievre committed:
Merge pull request #11337 from jolelievre/more-logs

Show more logs in admin

MathiasReker committed:
Unset Powered-By: PrestaShop

matks committed:
Merge pull request #11160 from rokaszygmantas/migration/import_improvements

Form handling improvements in import page step 1 and 2

PierreRambaud committed:
Merge pull request #11345 from eternoendless/fix-catalog

Update translations catalog

PierreRambaud committed:
Escape password render when installation is finished

Make sure base uri is a valid base path

jolelievre committed:
Add an rtlfix file to correctly hide the right side bar in rtl languages

Commit:
Fix missing dot in wording

Commit:
Remove invalid wordings

sarjon committed:
fixes crashing bulk action for some grids

sarjon committed:
build assets

sarjon committed:
remove unused include

PierreRambaud committed:
Merge pull request #11311 from mickaelandrieu/fix-regression-on-help-link

Be able to disable Help Link on Back Office pages

PierreRambaud committed:
Merge pull request #11384 from jolelievre/rtl-sidebar

Add an rtlfix file to correctly hide the right side bar in rtl languages

PierreRambaud committed:
Merge pull request #11385 from eternoendless/fix-wording

Fix missing dot in wording

sitte committed:
add total_shipping_tax_excl , incl in email data

>Is your feature request related to a problem? Please describe.
No way to use total shipping tax excl in order_conf email.

>Describe the solution you'd like
In Shop where everything is tax excluded its inconsistency

>Describe alternatives you've considered
overriding is not easy/imposible and total_shipping_tax_excl and total_shipping_tax_incl are right in class

sarjon committed:
build assets

PierreRambaud committed:
Merge pull request #10916 from sarjon/migrate/customers-list

Migrate customers listing

PierreRambaud committed:
Merge pull request #11357 from matks/improve-install-msg-for-beta-builds

Improve install warning message for beta builds

sitte committed:
Update PaymentModule.php

sarjon committed:
fix url to order view page

matks committed:
Fix img directory necessary for tests

sarjon committed:
fix escaped string encdoing

Commit:
Sort by locale if there is no access time

Commit:
Build assets

matks committed:
Merge branch 'develop' into categories-crud-tests-fixed

sarjon committed:
fix tests

sarjon committed:
add filter placeholders

Quetzacoalt91 committed:
Merge pull request #11394 from sarjon/fix-crashing-bulk-actions

Fix crashing for bulk actions

Quetzacoalt91 committed:
Merge pull request #11407 from PierreRambaud/module-manager-sorting

Module manager sorting

Quetzacoalt91 committed:
Merge pull request #11309 from jolelievre/import-loop

Add LegacyContext::getLegacyAdminLink

matks committed:
Fix conflict

Commit:
Regenerate Htaccess after saving domain informations

Commit:
Merge branch 'develop' of github.com:PrestaShop/PrestaShop into issue-9753

tomlev committed:
fix customer/address delete resulting in cart page

tomlev committed:
separate address delete in a specific function

MathiasReker committed:
Lossless compression

MathiasReker committed:
Fixed conflicts

MathiasReker committed:
Revert "Fixed conflicts"

MathiasReker committed:
Revert "fix conflicts"

This reverts commit c7eefa44c1a5911cf6855743696b60329341618c.

tomlev committed:
cs-fixer

Commit:
Move to dedicate function

Commit:
Noto Sans fonts updated to version 2.001 hinted

category: FO: theme classic, Noto Sans font update to 2.001
Fixes #11398 among the backtick bug on OS X. 
Exact font version: noto-source:20181019:f8f3770;ttfautohint (v1.8.2)
Conversion made with FontPlop version 1.2
Checked with Mac OS X that there are still 3.113 characters in Noto Sans-Regular after conversion.
Old size: 11 MiB, New size: 5.8 MiB

Commit:
Merge pull request #1 from ProBackup-nl/ProBackup-nl-patch-noto-sans

Noto Sans fonts updated to version 2.001 hinted

Quetzacoalt91 committed:
Merge pull request #11410 from PierreRambaud/fix/htaccess-11375

Changing primary domain name no longer reindexes simplified URLs

matks committed:
Build assets

Quetzacoalt91 committed:
Merge pull request #11416 from tomlev/issue-9782

fix customer/address delete resulting in error on cart page

Quetzacoalt91 committed:
Merge pull request #10199 from CaptainYouz/issue-9753

BO | Stock: Edit quantity placeholder

matks committed:
Merge branch 'develop' into categories-crud-tests-fixed

matks committed:
Rebuild assets

matks committed:
Merge pull request #11408 from matks/categories-crud-tests-fixed

Migrate Categories Add/Edit actions - tests fixed

Quetzacoalt91 committed:
Display configure button only if module is configurable

Quetzacoalt91 committed:
Merge pull request #11376 from PierreRambaud/missing-escapes

Some strings need to be escaped

manfield committed:
Merge branch 'ps_develop' into develop

manfield committed:
Code reduced

manfield committed:
Reviewed for linter compatibility

PierreRambaud committed:
Merge pull request #11421 from Quetzacoalt91/display-configure-if-configurable

Display configure button only if module is configurable

Quetzacoalt91 committed:
Merge pull request #11280 from hadrich-hatem/fix-test-1750buil4

Add some fixes to the script

Quetzacoalt91 committed:
Merge pull request #11417 from MathiasReker/optimize-images

Lossless compression of all images

mnastalski committed:
Improve backwards compatibility when setting an order's currency in Admin Orders

sarjon committed:
fix private note ui

sarjon committed:
update lists UI

dineshbadrukhiya committed:
Removed style to fix the bug on BO

sarjon committed:
make orders, carts, addresses and discounts list rows clickable

sarjon committed:
rename customer viewing command

Commit:
Merge pull request #11419 from ProBackup-nl/develop

Noto Sans fonts updated to version 2.001 hinted

Commit:
Merge pull request #11397 from sitte/patch-1

add total_shipping_tax_excl , incl in email data

sarjon committed:
fix key duplication in yaml

Quetzacoalt91 committed:
Merge pull request #11424 from dineshbadrukhiya/Fix-#11216

Improvement needs on hover of "Ok" button of source code modal on product page

Quetzacoalt91 committed:
Merge pull request #11235 from mnastalski/backend-orders-performance

Prevent unnecessary Order objects from being created in Admin Orders

PierreRambaud committed:
Merge branch '1.7.5.x' of github.com:PrestaShop/PrestaShop into merge2develop

Quetzacoalt91 committed:
Merge pull request #11456 from PierreRambaud/merge2develop

Merge 1.7.5.x in develop

sarjon committed:
build assets

sarjon committed:
fix link to customer address edit page

Quetzacoalt91 committed:
Fix SQL update when last_connection_date is null

sarjon committed:
cs fixer

PierreRambaud committed:
Merge pull request #11256 from manfield/develop

Fixes error reported in #9687, regarding carrier list in the modal to ...

PierreRambaud committed:
Merge pull request #11204 from andromaque/develop

[*] CO : Fixed incorrect comparison when out of stock

PierreRambaud committed:
Merge pull request #10839 from sarjon/migrate/languages-list

Migrate languages list

PierreRambaud committed:
Merge pull request #11362 from MathiasReker/header

Dont expose Powered-By: Prestashop

Quetzacoalt91 committed:
Merge pull request #11469 from Quetzacoalt91/upd-last-connexion-date

Fix SQL update when last_connection_date is null

sarjon committed:
build assets

Quetzacoalt91 committed:
Merge pull request #10992 from sarjon/migrate/customers-view

Migrate Customer view page

Quetzacoalt91 committed:
Merge pull request #9312 from fatmaBouchekoua/selenium/order_detail

Check order details (selenium tests)

Quetzacoalt91 committed:
Merge pull request #9309 from hadrich-hatem/button_header_product_page

Check buttons in header product page

PierreRambaud commented:
In fact the conflict is coming when it tries to merge your branch into develop branch.
You can use `git rebase` command or `git pull`

db306 committed:
Merge remote-tracking branch 'origin/delete_image' into delete_image

PierreRambaud commented:
Hum, look like something went wrong ^^
`Files changed 4,256 +216,018 −237,677`

PierreRambaud commented:
Can you please rebase your branch?

PierreRambaud reviewed (changes requested):
Waiting for rebase

marionf commented:
Hello @db306 

Since we have had no news for more than 30 days, I close this PR
Feel free to open a new one if needed
